### PR TITLE
PERF: slow tests

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #44624
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

Biggest difference is in test_get_loc_tuple_monotonic_above_size_cutoff going from 12.92 seconds t 4.72, all in improving DatetimeLikeArray.astype.

### CODE REVIEW

File: pandas/core/dtypes/missing.py [hunk 1]
```diff
@@ -644,3 +646,8 @@ def is_valid_na_for_dtype(obj, dtype: DtypeObj) -> bool:
 
     # fallback, default to allowing NaN, None, NA, NaT
     return not isinstance(obj, (np.datetime64, np.timedelta64, Decimal))
+
+
+def _is_string_dtype(dtype: np.dtype) -> bool:
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:651]: hmm can you simply move this to dtypes.core.common next to the other one

Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py]: Sure.  most recent commit also added an improvement for `random_text`

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**